### ISSUE TITLE
Структура БД

### ISSUE THREAD

**Issue by KateBushueva** [pf03/server#8]
зачем таблице `drafts` иметь поле `news_id` ? Насколько я понимаю, есть контент, который может быть или черновиком или новостью

**Comment by KateBushueva** [comment]
нигде не юзаются внешние ключи, хотя имхо это хороший инструмент для поддержания целостности. Ну и структурно сразу представляешь себе модель бд

**Comment by KateBushueva** [comment]
хотелось бы увидеть модель БД

**Comment by pf03** [comment]
> зачем таблице `drafts` иметь поле `news_id` ? Насколько я понимаю, есть контент, который может быть или черновиком или новостью

Да, контент может быть либо черновиком, либо новостью. Точнее контент это внутреннеe содержание черновика или новости. Цикл жизни новости следующий. 
Создание черновика `drafts/create`  - черновик создается, он еще не привязан к новости (news_id = null). 
Публикация новости админом `drafts/n/publish` -  черновик удаляется, пост создается. 
Редактирование новости `posts/n/edit` - новый черновик создается, уже привязанный к news_id. 
Публикация измененной новости админом `drafts/n/publish` - черновик удаляется, пост редактируется.

**Comment by pf03** [comment]
> нигде не юзаются внешние ключи, хотя имхо это хороший инструмент для поддержания целостности. 

Добавил внешние ключи. Все проверки связанных сущностей дублируются вручную. 
> Ну и структурно сразу представляешь себе модель бд
 
Собственно по названию столбцов все связи понятны

**Comment by pf03** [comment]
> хотелось бы увидеть модель БД

![db-structure](https://user-images.githubusercontent.com/15140589/129446179-9cd3ea6d-04a3-49ad-a4c2-0aa382046758.png)

**Comment by KateBushueva** [comment]
ок, по структуре:
- таблицы контент и фото: не нужно давать ссылки друг на друга (это избыточные связи) пусть будет в контенте ссылка на фото, а в таблице фото просто айди, описание (если оно там нужно) и, допустим, файл...
- по поводу постов и драфтов - по сути, это флаг, а не отдельные сущности.. контент может быть либо тем, либо другим но не одновременно. Поэтому я бы убрала эти таблицы и добавила бы просто поле (можно даже булевое типа `is_draft`) в таблицу контент.. Соответственно в таблице коммент будет ссылка не на пост-айди, а на контент-айди
- тэги - я бы сделала в контенте поле тэги (куда добавляем массив с айдишками тэгов) и таблицу айди: тэг - текст тэга

Вообще можешь почитать теорию про нормализацию таблиц в БД (приведение структуры бд к 3й нормальной форме, это, если не ошибаюсь, логическое проектировае БД)

**Comment by pf03** [comment]
> * таблицы контент и фото: не нужно давать ссылки друг на друга (это избыточные связи) пусть будет в контенте ссылка на фото, а в таблице фото просто айди, описание (если оно там нужно) и, допустим, файл...

Там нет избыточных связей. По заданию к посту привязано одно главное фото, и много дополнительных. Так вот в таблице contents название файла главного фото, в таблице photos название файлов дополнительных фото. Если я поменяю поле photo на main_photo это будет понятно? 
PS Сами фото в БД не хранятся, они находятся в папке photos и загружаются фронтендом по имени файла `localhost/photos/zvezdnoe_nebo_test.png`. На сервер загружаются в теле запроса `localhost/photos/upload?name=zvezdnoe_nebo_test.png`.
В папке `curl` есть тестовые запросы

**Comment by pf03** [comment]
> * тэги - я бы сделала в контенте поле тэги (куда добавляем массив с айдишками тэгов) и таблицу айди: тэг - текст тэга

Мой вариант ведь тоже верный, избыточных связей нет. https://qastart.by/class-6/90-svyazi-mezhdu-tablitsami-mnogie-ko-mnogim-many-to-many Это обязательно исправлять?

**Comment by pf03** [comment]
> * по поводу постов и драфтов - по сути, это флаг, а не отдельные сущности.. контент может быть либо тем, либо другим но не одновременно. Поэтому я бы убрала эти таблицы и добавила бы просто поле (можно даже булевое типа `is_draft`) в таблицу контент.. Соответственно в таблице коммент будет ссылка не на пост-айди, а на контент-айди

кмк так будет больше возможностей для ошибок. Например можно будет привязать коммент к черновику или черновик к черновику. В моей структуре такое невозможно, коммент и черновик можно привязать только к посту.
Но если надо, я исправлю конечно

**Comment by KateBushueva** [comment]
еще по структуре такой момент: используешь названия для филдов слова qsl (незарезервированные, но обладающие своей функциональностью) - типа `name` , `text` ... этого необходимо избегать

**Comment by KateBushueva** [comment]
> > * тэги - я бы сделала в контенте поле тэги (куда добавляем массив с айдишками тэгов) и таблицу айди: тэг - текст тэга
> 
> Мой вариант ведь тоже верный, избыточных связей нет. https://qastart.by/class-6/90-svyazi-mezhdu-tablitsami-mnogie-ko-mnogim-many-to-many Это обязательно исправлять?

тут спорный момент, ты добавляешь новую сущность, и размер это таблицы (связать каждый тэг с каждым контентом) будет крайне велик. В принципе, можно и оставить, но тогда у таблицы `tags_to_content` будет составной ключ, не нужно дополнительно в эту таблицу добавлять филд `id`

**Comment by KateBushueva** [comment]
> м нет избыточных связей. По заданию к посту привязано одно главное фото, и много дополнительных. Так вот в таблице contents название файла главного фото, в таблице photos название файлов дополнительных фото. Если я поменяю поле photo на main_photo это будет понятно?
> PS Сами фото в БД не хранятся, они находятся в папке photos и загружаются фронтендом по имени файла `localhost/photos/zvezdnoe_nebo_test.png`. На сервер загружаются в теле запроса `localhost/photos/upload?name=zvezdnoe_nebo_test.png`.
> В папке `curl` есть тестовые запро

В таком случае откуда ты берешь главное фото? это же тоже не просто ссылка на файл, там же есть название, какие-то метаданные

**Comment by KateBushueva** [comment]
> > * по поводу постов и драфтов - по сути, это флаг, а не отдельные сущности.. контент может быть либо тем, либо другим но не одновременно. Поэтому я бы убрала эти таблицы и добавила бы просто поле (можно даже булевое типа `is_draft`) в таблицу контент.. Соответственно в таблице коммент будет ссылка не на пост-айди, а на контент-айди
> 
> кмк так будет больше возможностей для ошибок. Например можно будет привязать коммент к черновику или черновик к черновику. В моей структуре такое невозможно, коммент и черновик можно привязать только к посту.
> Но если надо, я исправлю конечно

Ну у тебя есть бизнес-логика, которая не должна позволять привязать коммент к черновику. А черновик к черновику ты не привяжешь, объект контент будет только один. Этот момент я бы исправила.

**Comment by KateBushueva** [comment]
> Добавил внешние ключи. Все проверки связанных сущностей дублируются вручную.

я бы еще запретила удалять объекты, если на них идет ссылка из другого объекта

**Comment by pf03** [comment]
> еще по структуре такой момент: используешь названия для филдов слова qsl (незарезервированные, но обладающие своей функциональностью) - типа `name` , `text` ... этого необходимо избегать

исправил имена столбцов name, text, login
заодно прогнал все curl запросы и выявил несколько мелких косяков

**Comment by pf03** [comment]
> тут спорный момент, ты добавляешь новую сущность, и размер это таблицы (связать каждый тэг с каждым контентом) будет крайне велик. В принципе, можно и оставить, но тогда у таблицы `tags_to_content` будет составной ключ, не нужно дополнительно в эту таблицу добавлять филд `id`

добавил составной ключ и убрал `id`

**Comment by pf03** [comment]
> я бы еще запретила удалять объекты, если на них идет ссылка из другого объекта

Они и так не удаляются. Вот такие ошибки возникают при неправильном порядке удаления (пользователю они конечно не показываются, только в логах)

`UPDATE или DELETE в таблице "contents" нарушает ограничение внешнего ключа "photos_content_id_fkey" таблицы "photos"`

**Comment by pf03** [comment]
> В таком случае откуда ты берешь главное фото? это же тоже не просто ссылка на файл, там же есть название, какие-то метаданные

Ничего такого нет. В задании это не прописано, поэтому я сделал только ссылку на файл. 
Пользователь может загрузить фото на сервер, просмотреть фото по ссылке и просмотреть список всех фотографий на сервере (curl/photos)
Изменил contents.photo на main_photo

**Comment by KateBushueva** [comment]
тогда в чем смысл держать отдельную таблицу для фото? используй в контенте одну ссылку на фото как главное и массив ссылок на фото для тела. Просто когда ты создаешь таблицу для фото - подразумевается, что ты его рассматриваешь как объект. Поэтому возникает вопрос, почему у тебя фото в теле - это объекты, а главное фото - это ссылка на файл

**Comment by KateBushueva** [comment]
или делать все фото как объекты (ну типа для многоразового использования), но тогда главное фото тоже должно быть таким объектом

**Comment by pf03** [comment]
> используй в контенте одну ссылку на фото как главное и массив ссылок на фото для тела

Наверное я просто никогда не использовал массивы в бд, поэтому их избегаю) Я понял что нужно переделать. Убрать tagToContents, photos, drafts, posts. Чем меньше таблиц, тем меньше лишних связей и проверок, дублирования кода, и проще запросы. Это займет несколько дней.

**Comment by pf03** [comment]
Нет, все таки удалить tagsToContents не удастся. В таком случае таблица contents будет иметь массив ссылок на tags, а массив не может быть внешним ключом. 
Таблицу tags тоже никуда не уберешь, по условию теги имеют свои апи для создания, удаления и т.д.

**Comment by pf03** [comment]
Исправил по всем пунктам. Вот новая структура
![db-structure2](https://user-images.githubusercontent.com/15140589/130759315-b63925af-da4d-465b-99fc-93eb45be997d.png)

**Comment by KateBushueva** [comment]
а post_id в таблице contents сейчас куда ссылается? по коду - сам на себя (в смысле на `id` в таблице `content`), это как-то странно...

**Comment by KateBushueva** [comment]
и больше вопросов у меня нет по структуре

**Comment by pf03** [comment]
> а post_id в таблице contents сейчас куда ссылается? по коду - сам на себя (в смысле на `id` в таблице `content`), это как-то странно...

Да, так и есть. post_id это id поста, к которому привязан данный черновик. Если у нас в таблице не черновик, а пост, то post_id=null. Если черновик еще не был опубликован, то тоже post_id=null. Ты же сказала удалить таблицы posts и drafts, поэтому contents теперь ссылается на себя

**Comment by pf03** [comment]
Также и в таблице categories parent_id ссылается на себя, так что думаю это нормальная практика

**Comment by KateBushueva** [comment]
про категории понятно - ты мы выстраиваем иерархию категорий-подкатегорий. Кстати, вопрос: а что у тебя лежит в `parent_id` у  самой корневой категории?

**Comment by KateBushueva** [comment]
> Да, так и есть. post_id это id поста, к которому привязан данный черновик. Если у нас в таблице не черновик, а пост, то post_id=null. Если черновик еще не был опубликован, то тоже post_id=null. Ты же сказала удалить таблицы posts и drafts, поэтому contents теперь ссылается на себя

Немного не понимаю, а зачем? У тебя же есть флаг `is_draft` , который по логике и является показателем того, что перед нами черновик...

**Comment by pf03** [comment]
> Немного не понимаю, а зачем? У тебя же есть флаг `is_draft` , который по логике и является показателем того, что перед нами черновик...

Если новость публикуется в первый раз, то конечно незачем (post_id=null). Но у меня в проекте предусмотрено редактирование уже существующей новости. Тогда создается черновик, привязанный к post_id данной новости. Когда админ публикует новость, черновик удаляется, а пост обновляется. Я об этом писал ранее https://github.com/pf03/server/issues/8#issuecomment-898416557
